File: csharp-katas/solution/FeedParser/FeedParser.cs
namespace Katas.FeedParser;

using System.Globalization;

/// <summary>
/// The category of a malformed feed line. Validation short-circuits in a fixed
/// order (field-count → empty-symbol → bid → ask → qty), so at most one
/// <see cref="ErrorKind"/> is reported per line — the first rule that fails.
/// </summary>
public enum ErrorKind
{
    /// <summary>The line did not contain exactly four <c>|</c>-separated fields.</summary>
    WrongFieldCount,

    /// <summary>The symbol (first field) was empty.</summary>
    EmptySymbol,

    /// <summary>The bid (second field) was not a valid <see cref="double"/>.</summary>
    InvalidBid,

    /// <summary>The ask (third field) was not a valid <see cref="double"/>.</summary>
    InvalidAsk,

    /// <summary>The quantity (fourth field) was not a non-negative integer.</summary>
    InvalidQty,
}

/// <summary>
/// A parsed market-data quote: <c>SYMBOL|BID|ASK|QTY</c>.
/// </summary>
/// <param name="Symbol">The instrument identifier (the only field that allocates a string).</param>
/// <param name="Bid">The best bid price.</param>
/// <param name="Ask">The best ask price.</param>
/// <param name="Qty">The available size; always non-negative, hence <see cref="ulong"/>.</param>
public readonly record struct Quote(string Symbol, double Bid, double Ask, ulong Qty);

/// <summary>
/// A malformed line, tagged with its 1-based physical line number so an operator
/// can jump straight to the offending record in the raw feed.
/// </summary>
/// <param name="Line">The 1-based physical line number (blank/comment lines still count).</param>
/// <param name="Kind">Which validation rule failed.</param>
public readonly record struct ParseError(int Line, ErrorKind Kind);

/// <summary>
/// The outcome of one non-skipped feed line: exactly one of <see cref="Quote"/> or
/// <see cref="Error"/> is populated.
/// </summary>
/// <param name="Line">The 1-based physical line number this outcome came from.</param>
/// <param name="Quote">The parsed quote, or <c>null</c> if the line was malformed.</param>
/// <param name="Error">The parse error, or <c>null</c> if the line was a valid quote.</param>
public readonly record struct ParsedLine(int Line, Quote? Quote, ParseError? Error);

/// <summary>
/// A streaming, allocation-conscious parser for a pipe-delimited market-data feed
/// (<c>SYMBOL|BID|ASK|QTY</c> per line).
/// </summary>
/// <remarks>
/// <para>
/// <b>The zero-allocation story.</b> Each line is parsed through a
/// <see cref="ReadOnlySpan{T}"/> view over the source string. The four fields are
/// sliced with <see cref="MemoryExtensions.Split(ReadOnlySpan{char}, Span{Range}, char, StringSplitOptions)"/>,
/// which writes <see cref="Range"/> markers into a <c>stackalloc</c> buffer rather than
/// materialising a <c>string[]</c>. Numbers are parsed straight off the span via the
/// span-based <c>double.TryParse</c> / <c>ulong.TryParse</c> overloads, so no
/// intermediate substrings are allocated. The <em>only</em> heap allocation on the happy
/// path is the symbol string, produced with <c>ToString()</c> at the very end — because a
/// <see cref="Quote"/> must outlive the stack frame and a span cannot.
/// </para>
/// <para>
/// <b>TryParse over exceptions.</b> Malformed numbers are an expected, high-frequency
/// condition in a real feed. Exceptions for control flow would be catastrophic for
/// throughput, so we use the <c>TryParse</c> family and turn a <c>false</c> return into a
/// typed <see cref="ParseError"/>. A leading <c>'-'</c> makes <c>ulong.TryParse</c> return
/// <c>false</c>, which is exactly how the "quantity must be non-negative" rule is enforced.
/// </para>
/// <para>
/// <b>Streaming.</b> <see cref="Parse"/> is a lazy iterator (<c>yield return</c>): it pulls
/// one line at a time from the source <see cref="IEnumerable{T}"/> and emits one
/// <see cref="ParsedLine"/> per non-skipped line, so an unbounded feed can be processed with
/// O(1) memory. Because a <c>ref struct</c> span cannot live across a <c>yield</c>, the
/// per-line span work is isolated in a non-iterator helper that returns a plain value type.
/// </para>
/// <para>
/// <b>Line-numbered errors.</b> Every <see cref="ParseError"/> carries the 1-based
/// <em>physical</em> line number (blank and comment lines are counted even though they are
/// skipped), so diagnostics point at the exact source row.
/// </para>
/// <para>
/// <b>Invariant culture.</b> Prices and quantities are always parsed with
/// <see cref="CultureInfo.InvariantCulture"/> — a feed uses <c>'.'</c> as the decimal point
/// regardless of the host machine's locale, so parsing must never depend on ambient culture.
/// </para>
/// </remarks>
public static class FeedParser
{
    private const NumberStyles PriceStyles = NumberStyles.Float | NumberStyles.AllowThousands;
    private const NumberStyles QtyStyles = NumberStyles.Integer;

    /// <summary>
    /// Lazily parses <paramref name="lines"/>, yielding one <see cref="ParsedLine"/> for each
    /// line that is neither blank nor a comment.
    /// </summary>
    /// <param name="lines">The feed, one physical record per element.</param>
    /// <returns>
    /// A lazy sequence of parse outcomes. Blank lines and lines that (after trimming) start
    /// with <c>'#'</c> are skipped entirely — they produce no element but still advance the
    /// physical line counter.
    /// </returns>
    public static IEnumerable<ParsedLine> Parse(IEnumerable<string> lines)
    {
        var lineNumber = 0;
        foreach (var line in lines)
        {
            lineNumber++;
            if (ParseLine(line, lineNumber) is { } parsed)
            {
                yield return parsed;
            }
        }
    }

    /// <summary>
    /// Eagerly parses <paramref name="lines"/>, partitioning the outcomes into the successful
    /// quotes and the errors.
    /// </summary>
    /// <param name="lines">The feed, one physical record per element.</param>
    /// <returns>
    /// A tuple of (quotes, errors) in physical-line order. Convenience over
    /// <see cref="Parse"/> for callers that want the whole feed materialised.
    /// </returns>
    public static (IReadOnlyList<Quote> Quotes, IReadOnlyList<ParseError> Errors) ParseAll(IEnumerable<string> lines)
    {
        var quotes = new List<Quote>();
        var errors = new List<ParseError>();

        foreach (var parsed in Parse(lines))
        {
            if (parsed.Quote is { } quote)
            {
                quotes.Add(quote);
            }
            else if (parsed.Error is { } error)
            {
                errors.Add(error);
            }
        }

        return (quotes, errors);
    }

    /// <summary>
    /// Parses a single physical line into a <see cref="ParsedLine"/>, or <c>null</c> when the
    /// line is skipped (blank or comment). All span work is confined here so that
    /// <see cref="Parse"/> can remain an iterator.
    /// </summary>
    private static ParsedLine? ParseLine(string line, int lineNumber)
    {
        var span = line.AsSpan().Trim();

        // Blank or comment lines are skipped — not a record, not an error.
        if (span.IsEmpty || span[0] == '#')
        {
            return null;
        }

        // A buffer of 5 ranges lets us distinguish "exactly 4" from "5 or more":
        // Split fills at most destination.Length ranges, packing any overflow into the last.
        Span<Range> ranges = stackalloc Range[5];
        var count = span.Split(ranges, '|');
        if (count != 4)
        {
            return Error(lineNumber, ErrorKind.WrongFieldCount);
        }

        var symbol = span[ranges[0]];
        if (symbol.IsEmpty)
        {
            return Error(lineNumber, ErrorKind.EmptySymbol);
        }

        if (!double.TryParse(span[ranges[1]], PriceStyles, CultureInfo.InvariantCulture, out var bid))
        {
            return Error(lineNumber, ErrorKind.InvalidBid);
        }

        if (!double.TryParse(span[ranges[2]], PriceStyles, CultureInfo.InvariantCulture, out var ask))
        {
            return Error(lineNumber, ErrorKind.InvalidAsk);
        }

        // ulong.TryParse rejects a leading '-', enforcing the "non-negative" rule for free.
        if (!ulong.TryParse(span[ranges[3]], QtyStyles, CultureInfo.InvariantCulture, out var qty))
        {
            return Error(lineNumber, ErrorKind.InvalidQty);
        }

        // Only now do we allocate — the symbol string must outlive the stack frame.
        return new ParsedLine(lineNumber, new Quote(symbol.ToString(), bid, ask, qty), null);
    }

    private static ParsedLine Error(int lineNumber, ErrorKind kind) =>
        new(lineNumber, null, new ParseError(lineNumber, kind));
}

File: java-katas/solution/src/main/java/org/kata/feedparser/FeedParser.java
package org.kata.feedparser;

import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.atomic.AtomicInteger;
import java.util.stream.Stream;

/**
 * Parses a streaming market-data feed of pipe-delimited records into typed {@link Quote}s, lazily,
 * reporting each malformed line with its physical position rather than throwing.
 *
 * <h2>The feed format</h2>
 * One record per line: {@code SYMBOL|BID|ASK|QTY}. Two line kinds are <b>skipped</b> (they are not
 * records and not errors): after trimming, a <b>blank</b> line, or one that <b>starts with
 * {@code #}</b> (a comment). Skipped lines still advance the physical line counter — see below.
 *
 * <h2>Why lazy</h2>
 * The public {@link #parse(Stream)} maps an input {@code Stream<String>} of lines to a
 * {@code Stream<ParsedLine>} without materialising either. That matters for a real feed: it may be
 * enormous (a day's ticks) or effectively unbounded (a live socket adapted to a {@code Stream}), so
 * buffering the whole thing into a {@code List} first is a non-starter. A lazy pipeline processes one
 * line at a time and lets the terminal operation decide how much to pull — {@code findFirst},
 * {@code limit}, a short-circuiting search, or a full drain all work without changing the parser.
 *
 * <h2>Threading the physical line number</h2>
 * Every {@link ParseError} carries the <b>1-based physical line number</b> — counting <i>every</i>
 * input line, including the blank/comment lines that are themselves skipped. An operator diagnosing
 * a bad tick reads the raw file; "line 6" must mean the sixth line of that file, not the sixth
 * successfully-parsed record. Because {@link Stream} has no built-in element index, we thread the
 * counter through with an {@link AtomicInteger} incremented once per input line inside the mapping
 * step. This stays correct under a lazy, ordered, sequential pipeline (one increment per element, in
 * encounter order); it would <i>not</i> be safe to parallelise, which is exactly why {@link #parse}
 * returns an ordered sequential stream and never calls {@code parallel()}.
 *
 * <h2>Per-line outcome as a result type</h2>
 * Each non-skipped line becomes a {@link ParsedLine}: {@link ParsedLine.Ok} (a {@link Quote} plus its
 * line) or {@link ParsedLine.Err} (a {@link ParseError}). Errors are in-band values, not exceptions —
 * a single malformed line must never abort the stream and lose the good ticks behind it. The sealed
 * {@code ParsedLine} lets callers handle both arms exhaustively (the compiler enforces it).
 *
 * <h2>Validation order (short-circuits on first failure)</h2>
 * A line is validated in a fixed order, reporting only the first problem:
 * <ol>
 *   <li>exactly four fields → else {@link ErrorKind#WRONG_FIELD_COUNT};</li>
 *   <li>non-empty symbol → else {@link ErrorKind#EMPTY_SYMBOL};</li>
 *   <li>bid parses as {@code double} → else {@link ErrorKind#INVALID_BID};</li>
 *   <li>ask parses as {@code double} → else {@link ErrorKind#INVALID_ASK};</li>
 *   <li>qty parses as a <b>non-negative</b> {@code long} → else {@link ErrorKind#INVALID_QTY}.</li>
 * </ol>
 *
 * <h2>The {@code split("\\|", -1)} gotcha</h2>
 * Field-splitting uses {@link String#split(String, int)} with a limit of {@code -1}. The default
 * {@code split(regex)} (limit 0) <b>discards trailing empty strings</b>: {@code "TOO|1.0|2.0|".split("\\|")}
 * yields {@code ["TOO","1.0","2.0"]} — length 3, silently swallowing a genuinely malformed trailing
 * field and mis-reporting it as {@code WRONG_FIELD_COUNT} for the wrong reason (or worse, matching 4
 * by accident on other inputs). Limit {@code -1} keeps every empty field, so the field count is exact
 * and empty fields are surfaced (e.g. an empty symbol is preserved rather than dropped). The
 * {@code |} is also regex-escaped ({@code "\\|"}) because a bare {@code |} is regex alternation.
 */
public final class FeedParser {

    private FeedParser() {
    }

    /**
     * Lazily maps a stream of raw feed lines to a stream of per-line outcomes.
     *
     * <p>Blank and {@code #}-comment lines (after trimming) produce <i>nothing</i> — they are
     * filtered out via {@code mapMulti}, which emits zero-or-one {@link ParsedLine} per input line
     * while keeping the pipeline lazy. Each surviving line becomes exactly one {@link ParsedLine.Ok}
     * or {@link ParsedLine.Err}. The returned stream is ordered and sequential; the physical line
     * counter it threads is only correct under that guarantee.
     */
    public static Stream<ParsedLine> parse(Stream<String> lines) {
        // One increment per input line, in encounter order. Correct only because the pipeline is
        // ordered + sequential (see class Javadoc) — do not parallelise this stream.
        AtomicInteger lineNo = new AtomicInteger(0);
        return lines.<ParsedLine>mapMulti((raw, downstream) -> {
            int line = lineNo.incrementAndGet();
            String trimmed = raw.trim();
            // Skip rule: blank or comment lines are neither records nor errors — emit nothing,
            // but the counter has already advanced so later errors keep their physical position.
            if (trimmed.isEmpty() || trimmed.startsWith("#")) return;
            downstream.accept(parseLine(trimmed, line));
        });
    }

    /**
     * Eagerly drains the feed, partitioning outcomes into the list of good {@link Quote}s and the
     * list of {@link ParseError}s (both in feed order).
     *
     * <p>A convenience terminal for the common "give me everything" case, built on the lazy
     * {@link #parse(Stream)} so there is a single source of truth for the parsing rules.
     */
    public static ParseSummary parseAll(Stream<String> lines) {
        List<Quote> quotes = new ArrayList<>();
        List<ParseError> errors = new ArrayList<>();
        parse(lines).forEach(parsed -> {
            // Exhaustive over the sealed ParsedLine — the compiler rejects a missing arm.
            switch (parsed) {
                case ParsedLine.Ok ok -> quotes.add(ok.quote());
                case ParsedLine.Err err -> errors.add(err.error());
            }
        });
        return new ParseSummary(quotes, errors);
    }

    /**
     * Validates and coerces one already-trimmed, non-skipped line into a {@link ParsedLine},
     * applying the fixed validation order and short-circuiting on the first failure.
     */
    private static ParsedLine parseLine(String trimmed, int line) {
        // Limit -1 keeps trailing empty fields so the count and empty-symbol checks are exact.
        String[] fields = trimmed.split("\\|", -1);
        if (fields.length != 4) return err(line, ErrorKind.WRONG_FIELD_COUNT);

        String symbol = fields[0];
        if (symbol.isEmpty()) return err(line, ErrorKind.EMPTY_SYMBOL);

        double bid;
        try {
            bid = Double.parseDouble(fields[1]);
        } catch (NumberFormatException e) {
            return err(line, ErrorKind.INVALID_BID);
        }

        double ask;
        try {
            ask = Double.parseDouble(fields[2]);
        } catch (NumberFormatException e) {
            return err(line, ErrorKind.INVALID_ASK);
        }

        long qty;
        try {
            qty = Long.parseLong(fields[3]);
        } catch (NumberFormatException e) {
            return err(line, ErrorKind.INVALID_QTY);
        }
        // Qty is a size; a negative size is meaningless. Long.parseLong happily accepts "-5",
        // so the non-negative constraint is a separate, explicit guard.
        if (qty < 0) return err(line, ErrorKind.INVALID_QTY);

        return new ParsedLine.Ok(line, new Quote(symbol, bid, ask, qty));
    }

    private static ParsedLine err(int line, ErrorKind kind) {
        return new ParsedLine.Err(new ParseError(line, kind));
    }

    /**
     * The result of draining an entire feed: every good {@link Quote} and every {@link ParseError},
     * each list in feed order. A {@code record} for value-equality so a test can assert on the whole
     * summary at once.
     */
    public record ParseSummary(List<Quote> quotes, List<ParseError> errors) {
    }
}
